### DISCUSSION TITLE
Where do you see yourself in 5 years?

### DISCUSSION THREAD

**Discussion by Westbrook** [w3c/webcomponents-cg#39] [Ideas]
It's always good to have a plan, a vision, a North Star (if you will), and that includes Web Components. In association with our API/Specs Status Report for 2022, we're looking to include a rough visions for Web Component Specs in the next five years and how browser implementers can support getting there along a path that is successful for all involved in the process. To do so, we want to know where you think Web Components APIs and Specs should be in the next five years.

There's no guarantee any of this comes to life, there's not even a reasonable way to include that all of your ideas get included in the report, but I can guarantee that we'll all be excited to hear your thoughts. From these ideas we'll formulate a general five year vision of Web Components and work back from there to outline APIs and specs that need to solidify and land to make these futures, or a future like them, a reality.

Thanks in advance for sharing your thoughts!

**Comment by Westbrook** [comment]
# Accessibility
- A shadow root should be able to clearly define the relationship that is maintains between its parent context and the content within it in order to appropriately construct the accessibility tree of the content that it represents.
# Selectors
- A custom element should be able to style itself taking its context in a parent DOM tree into account. This doesn't need to pierce to the root of the application, but being able to style on context, similar to `:host-context()` which was a cool thought, would open a number of avenues along the lines of `@container style(--thing: 'thing')` in the case the this approach should leverage more esoteric selecting to make it possible.
- A custom element should be able to style the dependents of its slotted content. The is particularly necessary when leverage combinator built-ins like `<table>`, `<ul>`, or `<details>`.
- A custom element should be able to apply styles to slotted content at an `@layer` context/specificity higher than "user agent".
- A custom element should be able to style itself or its children on the present of content (text nodes or DOM) in its slots.
- A custom element should be able to style in `:nth-*` precision on content in a specific slot.
# Stylesheets
- CSS Module scripts should be available everywhere.
- A Stylesheet object whether achieved via construction, asserted import, `<style>` tag, or `<link>` tag should be applicable to the `adoptedStyleSheets` property on a DOM tree.
- A primitive by which multiple Stylesheet objects can be "bundled" without the requirement of a JS interchange file should be created so that asserted imports can be machined together as needed to support better load performance.
- Stylesheet objects should have an identifier that allows them to be adopted declaratively without needing to repeat the same style rules multiple times in a DOM tree.
- The stylesheet object should be enhanced to allow for selectors they contain to be referenced in a JS templating context.
# Style customization
- There should be a unifying API by which all of the elements in an application can be styled without having to know the full complexity of the DOM tree. Something opt-in for component developers and building off of previous attempts (e.g. `@apply`) or current capabilities (e.g. `::part()`) to unify the way you point to a known system without devolving into massive Custom Property management schemes.
# Defining/Upgrading
- Custom elements should be able to be defined lazily.
- Custom elements should be able to be undefined/redefined.
- A mechanism for upgrading lazily should also be surfaced.
- A declarative technique for defining custom elements should be surfaced.
- A custom element import assertion should allow for defining an element from a single file.
# Description
- A declarative technique for describing a custom element without the need for a JS class should be surfaced.
  - To include things like:
    - tag name registration
    - templating
    - observed attribute and callback declaration
    - management of property reactivity
  - Known issue: this may never be the "optimal" way to write a custom element, but being able to machine down to it will be important for no JS context while there are not enough HTML primitives for fully define the complexities of modern day web applications.
# Templating
- A declarative method for template DOM should be created.
- Either through the HTML assertion or a new one developers should be able to import template files into the JS module graph
- Templates should have an identifier that allows them to be adopted declaratively without needing to repeat the same DOM structure multiple times in a DOM tree.
- Templating API should be low level enough to customize the end user "language" in which they are leveraged, but also high level enough to surface a fast path "language" for general use.
# Module graph
- Asserted imports should participate in the JS module graph
- Asserted imports of either HTML or custom element type should automatically "scope" the coding context that the imported file represents, including the custom element registry present in that file.
- declaratively created, tag linked, and asserted imports should be able to share in the deduplication inherent in the JS module graph regardless of the origin of the file's inclusion.
- `<link rel="modulepreload">` or equivelent should be available for all members of the module graph.

**Reply by Westbrook** [reply]
# Defining/Upgrading (addendum)

- Custom element need a method for defining themselves in a "long lived" manner. This would allow developers to define custom elements to a cache managed at something like the domain level such that the definition is available across multiple page loads.

**Comment by dgp1130** [comment]
Not sure if this is totally what you're looking for, but two related areas I'd like to see explored more which haven't been mentioned yet:

# Server side rendering

Personally, I would love to see some improvements in the SSR'd web components space. The web ecosystem has been moving more and more towards SSR infrastructure again, and a framework-independent specification for an SSR-capable component could be just as valuable on the server as it is on the client. I have no idea what this would look like, especially given that it's not really a browser problem, but I would really like to see more experimentation and standardization in this area. Some specific items (most of which inspired by [this thread](https://github.com/webcomponents-cg/community-protocols/issues/7#issuecomment-1134001778)):

* Web components should be able to support server side rendering in some standardized capacity.
* Multiple declarative shadow DOM elements should be able to share the same stylesheet.
    * Today you either need to duplicate inline `<style />` tags or use a `<link />` which causes a FOUC.
* Web components should have a means of declaring/linking resources needed at runtime on the client (JS, CSS, images, fonts, etc.) to support static site generation use cases and bundlers.
* SSR'd web components should be able to tree-shake / drop their client-side implementation if they don't require CSR (ie. support *SSR-only* web components).
* Web devs should be able to rely on well-defined primitives and contracts to defer / dynamically hydrate components.

My end goal here is to be able to take existing web components in a standardized format and render them on the server, providing a common format for components to be used across any server framework / infrastructure. This would enable a number of performance and usability improvements as well as unlock a whole new environment for web component authors to leverage without relying on frameworks or special support in applications. Imagine a full design system which is completely framework-agnostic but also supports SSR, with most components having zero-JS runtime since all the work is done on the server!

# "Webcomponents-over-the-wire"

A recent idea I've been exploring is a more [standards-driven HTML-over-the-wire approach](https://blog.dwac.dev/posts/html-fragments/), powered by web components. The TL;DR is that I would love to have a supported way for an SSR'd web component to be an API response type. Instead of returning a JSON response that gets CSR'd into the DOM, let's just return an SSR'd component which just gets attached to the DOM. This pulls on many of the above SSR issues but also brings up some other challenges.

(Full disclosure, I'm not totally convinced all of these are good ideas, but as a long shot "North Star" vision of what web development could look like I think it's worth at least mentioning them in this context.)

* JS should be able to import an SSR'd web component as an HTML module.
* JS should be able to import the same HTML module multiple times with different SSR'd content.
* HTML modules should be CSP/TrustedTypes supported.
* Browsers should support an [`<embed />`](https://blog.dwac.dev/posts/html-fragments/#ecosystem) tag.

My vision for this is being able to write an HTML-over-the-wire application entirely based on web components and existing standards without any special tooling, infrastructure, or heavy coupling to a server framework. This would give developers the flexibility to easily choose where rendering happens (client or server) as well as mix and match these approaches with different components and use cases in the same application. This would also all be possible without overly committing their application's architecture to SSR or CSR, letting them choose the best tool for the job at any given time.

Again, I'm not totally sure this is what you're looking for in this thread, but hopefully these are at least interesting ideas to consider in a "North Star" vision. :grin:

**Reply by Westbrook** [reply]
Great start! Thanks. 🚀 

Already some productive overlap 😅. Sussing out those "a part of all the visions" parts _seems_ like a good direction to outline?

Couple of things I'd love to hear more about:
- "SSR'd web components should be able to tree-shake..."
  - This (and a lot of your other SSR points) feel very _server_-centric. I wonder if the w3c has ever leaned into "server specs"? All the same, has this idea of a "won't need CSR" custom element been discussed in the context of the Element Renderer protocol that the Lit team is interested in pushing
- "JS should be able to import an SSR'd web component as an HTML module."
  - If you are already running JS, why SSR?
  - What's more, what does SSR "a web component" mean for you? 
  - Does the current DSD API in Chrome allow us to construct and upgrade/append DOM that has DSD in it? If it could, would that be similar to what you're thinking here?
- "JS should be able to import the same HTML module multiple times with different SSR'd content."
  - Are there other types of files that allow for this today? Like if you request a JS module later in the page lifecycle it returns the deduplicated file, not a new file. I guess CSS doesn't actually run deduplication, so with the expiries managed you could get this just due to time passing? Usually, how would a server know to return a different file? Should this be managed with URL params?
-  `<embed />` would be a new "non-custom" element, right? Have you brought this up to any of the HTML/DOM working groups, or maybe to open-ui.org? Seems cool!

**Reply by dgp1130** [reply]
> * "SSR'd web components should be able to tree-shake..."
>   
>   * This (and a lot of your other SSR points) feel very _server_-centric. I wonder if the w3c has ever leaned into "server specs"? All the same, has this idea of a "won't need CSR" custom element been discussed in the context of the Element Renderer protocol that the Lit team is interested in pushing

They are server-centric and I agree that's awkward from a standardization perspective. HTTP does cover servers and Node folks are involved in the standards process from my understanding, so this isn't unprecedented even if it's a bit weird. I'm not familiar enough with the different standards groups to know exactly where this would fit.

It's possible this makes more sense as a community protocol or some other mechanism, which is why I'm framing this more as "I want a way to author and use components in a standardized fashion which can use SSR but aren't dependent on any particular framework". Maybe that's more of a problem space than a vision of a desired solution.

> * "JS should be able to import an SSR'd web component as an HTML module."
>   
>   * If you are already running JS, why SSR?
>   * What's more, what does SSR "a web component" mean for you?
>   * Does the current DSD API in Chrome allow us to construct and upgrade/append DOM that has DSD in it? If it could, would that be similar to what you're thinking here?

SSR can be useful outside of initial page load contexts. If a lot of work is done to compute the rendered DOM or you need to make a server round trip anyways, why bother implementing and supporting a CSR implementation of a component if all your use cases can actually be met via SSR. It also means less JS is required at runtime if all it does is incrementally update an already rendered element instead of rerendering the whole thing from scratch.

This is explained more in the [blog post](https://dwac.dev/posts/html-fragments/) I linked about SSR-ing specific pieces of DOM on-demand and using custom elements to hydrate it with functionality in an "islands-style" architecture.

I'm imagining an "SSR-ed web component" as a custom element with pre-rendered children. Something like:

```html
<my-tweet>
  <template shadowroot="open">
    <style>...</style>
    <span>Hello from an SSR tweet!</span>
    <button>Like</button>
  </template>
  <script src="my-tweet.js” type="module"></script>
</my-tweet>
```

You can do this in a typical web page today, but with HTML modules this could be a complete response to `GET /api/tweet?id=1234`. Importing this would be a complete SSR'd web component ready to attach to the DOM. Since the server did the work of rendering the component, all the client side JS actually needs to do is bind an event listener for the like button, no need for a CSR framework or library to do this ergonomically.

> * "JS should be able to import the same HTML module multiple times with different SSR'd content."
>   
>   * Are there other types of files that allow for this today? Like if you request a JS module later in the page lifecycle it returns the deduplicated file, not a new file. I guess CSS doesn't actually run deduplication, so with the expiries managed you could get this just due to time passing? Usually, how would a server know to return a different file? Should this be managed with URL params?

To clarify what I mean by "same HTML module with different content", I mean a single component definition with different SSR'd content. For example if I request `/api/tweet?id=1234` and then request `/api/tweet?id=4321`, I expect to see the same component with different content:

```html
<my-tweet>Hello world from tweet 1234!</my-tweet>
<!-- vs. -->
<my-tweet>Hello again from tweet 4321!</my-tweet>
```

The definition of `my-tweet` and the associated JS and CSS would be unchanged, but the SSR'd content could be different.

This gets extra confusing if you consider that a tweet could be edited, so requesting `/api/tweet?id=1234` multiple times might give different responses. The modules themselves might used different paths or have a cache busting query param, but use the same components, scripts, and styles.

See https://blog.dwac.dev/posts/html-fragments/#code-vs-content for more details here.

> * `<embed />` would be a new "non-custom" element, right? Have you brought this up to any of the HTML/DOM working groups, or maybe to open-ui.org? Seems cool!

I haven't been very involved with the standards process, so I've only really brought this up in the blog post and some Twitter threads. I agree it's not really a custom element, so this might not be the right venue for it, but it's usage is heavily dependent on the "webcomponents-over-the-wire" approach I'm suggesting here, so I think it definitely fits with this North Star vision, even if this working group might not be the one whoch actually lands the proposal.

**Comment by sashafirsov** [comment]
# Unified resource reference and importmaps
Whether it is a `template`, or custom element defined declaratively, or data island, those need to be referenced as in current root scope as in page level as in external page/within external page. 

[URL](https://developer.mozilla.org/en-US/docs/Web/API/URL/URL) gives ability to refer the page DOM element, external resource and partially location within external resource. Unfortunately it is not coupled with importmaps to make externally served libs viable. 

In addition to generic ability, the attributes like `src`, `target`, etc. need to be importmaps plus in-resource treated. 

This concept is covered by several web component libs and a must feature for Web Components driven web app MVP.

**Comment by bennypowers** [comment]
If I could sum it up it one phrase it would be "HTML Apps". In five years' time, web components and associated technologies should be sufficiently mature to allow simple web apps to be developed little or no userland javascript, and complex apps with far less than is the norm today.

To get there we'll need at least
- HTML Imports
- Template Parts
- Declarative Custom Elements

And preferably something like declarative ElementInternals (aka `is` :stuck_out_tongue:)

**Reply by sashafirsov** [reply]
For "HTML Apps" in addition to list ^^ the data access layer (DAL) is needed as well. Which includes
* declarative data definition,
* binding data to DOM

Concept is quite popular in existing libs, just declarative syntax is not polished enough. `iron-ajax` paved some path. Template slots is one of way for binding the data to component and could be evolved into transformation pipe.
The `Parts` proposal gives the inclusion (kind of binding) but missing the data definition.

**Comment by thescientist13** [comment]
# Full Stack Web Components

With the advent of ESM on the server, supplemented by the rapid rise of web first serverless and edge runtimes backed by the [**WINTER CG**](https://wintercg.org/), what I am most excited for is the potential of _a web standards based development paradigm from server to browser for Web Components._

Through a combination of ESM, `fetch` / `Request` / `Response`, `URL`, Declarative Shadow DOM, and Import Assertions, what can be done on the web and on the server is becoming even more aligned than ever, and I think this can be an area for Web Components to really shine, where we can have our cake (write JS) and eat it too (ship more HTML / less JS).  🍰 

## Templating

Page and layout templating using components, common in full stack JS frameworks like Next and SvelteKit can be achieved with just a little bit of [work](https://github.com/ProjectEvergreen/wcc/blob/master/src/dom-shim.js) / [community alignment](https://github.com/webcomponents-cg/community-protocols/issues/35), and can make the server DX seamless.  Not that I have anything wrong with templating languages, but between this and template literals, I am totally ready to never have to use some bespoke DSL again. 

```js
import './components/footer.js';
import './components/header.js';

export default Layout extends HTMLElement {
  connectedCallback() {
    this.innerHTML = `
      <app-header></app-header>

      <h1>Welcome To My Website!</h1>

      <app-footer></app-footer>
    `
  }
}
```

> _I have been exploring this scenario already in my project [**Web Components Compiler**](https://github.com/ProjectEvergreen/wcc), and have been liking this pattern a lot, even running it on [the edge](https://github.com/thescientist13/web-components-at-the-edge)!  It was also designed to not ship that layout JS too, unless you need it at runtime._

## Over The Wire

Was chatting about this one recently with @dgp1130 and have some thoughts about how current and future horizon specs could leverage the constituent parts of a component definition (HTML / CSS / JS) independently.  This would support being able to optimize the workload for the given environment the component is running in (e.g. web vs server).  Thinking a combination of [**DOM Parts**](https://github.com/WICG/webcomponents/blob/gh-pages/proposals/DOM-Parts.md) and / or [**HTML Includes**](https://github.com/whatwg/html/issues/2791) and / or even [**Import Assertions**](https://github.com/tc39/proposal-import-assertions) could shake something like this loose?

For example, on the server where you may just care about returning static HTML, the below is a bike shed at what something like DOM Parts with HTML templates could look like.

_Template_
```html
<product-details>
  <h1>${name}</h1>
  <h2>${description}</h2>
  <p>Price: ${price}</p>
  <img src="${image}" alt="${name}"/>
  <button data-product-id="${id}">Buy Now</button>
</product-details>
```

_Handler Function_
```js
import detailsTemplate from './components/product-details/product-details.html'; 

export default async function handler (req, res) {
  const data = await fetch(`http://example.com/api/products/id=${req.id}`).then(resp => resp.json());
  const template = new Template(detailsTemplate, data);

  res.send(template.getContents());
}
```

But on the client, where say that product details can be interactive and you want to attach behavior (event handling), you'll want the JS there too, and of course still want to use that same HTML template for hydration (Declarative Shadow DOM).
_Custom Element_
```js
// or maybe globally loaded via an HTML Include <link rel="html" href="/components/product-details/product-details.html"> ???
import detailsTemplate from 'http://cdn.example.com/components/product-details/product-details.html';

const template = document.createElement('template');
template.innerHTML = detailsTemplate;

class ProductDetails extends HTMLElement {
  connectedCallback() {
    if (!this.shadowRoot) {
      this.attachShadow({ mode: 'open' });
      this.shadowRoot.appendChild(template.content.cloneNode(true));
    } else {
      this.shadowRoot.querySelector('button').addEventListener(this.buyProduct);
    }
  }

  buyProduct() {
    const id = this.shadowRoot.querySelector('button').getAttribute('data-product-id');
    prompt(`Do you want to buy ${id}?`);
  }
}

customElements.define('product-details', ProductDetails);
```

One advantage of this system here is no need for a DOM shim on the server side, but still have a single source of truth for your component definition across frontend and backend.

## Styling

Per @Westbrook, a good story around CSS would be vital as well.  I think import assertions would work really well here too, and / or in the interim, custom loader hooks like with NodeJS's `--experimental-loader` flag. 

```js
import detailsTemplate from './product-details/product-details.html';

import detailsStyles from './product-details/product-details.css' assert { type: 'css' }; // import assertion example
import detailsStyles from './product-details/product-details.css'; // or custom loader example

const template = document.createElement('template');

template.innerHTML = `
  <style>
    ${detailsStyles}
  </style>
  ${detailsTemplate}
`;

class ProductDetails extends HTMLElement {
  ...
}

customElements.define('product-details', ProductDetails);
```

> _I'm currently exploring feasibility with the `--experimental-loader` flag and have had good success so far, like playing around with [compiling JSX down to Web Components](https://merry-caramel-524e61.netlify.app/examples/#jsx)!_  🤨 

----

I think being able to facilitate these kind of developer workflows natively all through the runtime(s) is pretty compelling IMO.  Just install NodeJS and / or opening your browser and wherever you find yourself, you can use that Web Component! :green_heart:

All in all, it's never been a better to time to bet on the web and I would love to see (and help) Web Components lead the charge into this future! ✌️

**Reply by sashafirsov** [reply]
missed connection between `template` and `ProductDetails`.

With DCE
```html
<define tag="product-details" template="./product/templates.html#details" /> <!-- css loaded from within -->
```
or JS with extended HTMLElement 
```js
class ProductDetails extends HTMLElement {
  // import returns Promise<TemplateHTMLElement>, css loaded from within HTML with relative URL, 
  // hydration driven module loading
  static get template(){ return import('./product-details/product-details.html', {as:TemplateHTMLElement} ); } 
  ...
}
```

**Comment by sashafirsov** [comment]
# templating
* available from same document DOM when provided
* load from external document (DOM), loaded on demand. A library of templates/templates module.
* from html module ( ideally would match case ^^ but seems proposal goes "wrong" direction towards string presentation)
* re-defined in runtime(dynamic templating) by innerHTML/custom DOM fragment.

Externally defined templates source document have to be reused without the need to be injected directly into current document. When the page changes, templates DOM remain in memory and ready to be reused or offloaded.

**Comment by dmitriid** [comment]
5 years is not nearly a long-term planning. They turn *11 years* this year (starting with the original poposal by Alex Russel).

What we see in the list is "how can we patch the hundreds of holes in web components that no one stopped to think about, and instead just barrelled through with implementations".

What the web *really* needs in the next 5 years is:

- stop
- clearly define the actual vision and end goal for web components
- and only *then* continue

All the proposals seem to be just a continuation of "hell yeah, we know *what* to do" without ever, *ever* asking "why do we do this"? All of the issues are the issues web components themselves created. It's a self-perpetuating source of constant work and churn.

**Comment by jaredcwhite** [comment]
# Isomorphic Template Syntax That's Compatible with DSD

(First of all, great ideas and thoughts shared thus far!)

The biggest issue for me by far right now is the lack of any sort of template specification that is server-language agnostic (aka doesn't require the use of JavaScript) _and_ is compatible with DSD.

Let me explain. We can use the ubiquitous "counter" component example because it so easily describes the issue.

Let's say I want the counter to start at 1. So I can start to write a component like this:

```html
<my-counter count="1"></my-counter>
```

OK, so far so good.

Now I want to add a button in the shadow DOM to increment the counter as well as display the counter:

```html
<output></output>
<button>Increment</button>
```

Now I have two questions. (1) How does the counter value get injected into `<output>`? (2) How do I attach an event listener to `<button>`?

Maybe we can punt on the event listener problem. Do it imperatively with a JS statement in `connectedCallback`…

Still, we definitely have the issue of output. OK no problem, we use some sort of simple Handlebars-like syntax.

```html
<output>{{ count }}</output>
<button>Increment</button>
```

We can render that into a DSD:

```html
<template shadowroot="open">
  <output>1</output>
  <button>Increment</button>
</template>
```

OK, but now once my component has been upgraded on the client and a button click or an attribute or prop changes `count` to `2`, how do I know that `output` was previously bound to `count` so it should update from 1 to 2?

Well we'd likely have a copy of the template stored client-side with the JS. But now we have to start doing DOM diffing between the template and the DSDed output. It has to "figure out" that `<output>1</output>` started out life as `<output>{{ count }}</output>`. Hmm.

In my view, the answer is clear. We shouldn't use something like Handlebars. We should use something like Vue, Knockout, and other frameworks which can bind things via attributes. And then let those template bindings flow unchanged through the DSD rendering process. Let's revisit. In pseudocode:

```html
<output bind="count"></output>
<button>Increment</button>
```

That renders out as:

```html
<template shadowroot="open">
  <output bind="count">1</output>
  <button>Increment</button>
</template>
```

Aha! Right there in the DSD, the template syntax has flowed from server to client! Now if the value of `count` changes, it can change the content of `<output>` because it can find that binding.

OK, but what about the other direction? Can't we do something similar with event handlers? Again, looking to syntax like Vue's (or Lit's for that matter) for inspiration here. Not sure if it's actually valid HTML, but browsers don't seem to mind.

```html
<template shadowroot="open">
  <output bind="count">1</output>
  <button @click="increment">Increment</button>
</template>
```

It turns out that's actually the _easiest_ problem to solve because we likely don't care about event handling on the server. We just want to pass some indication along to the client that it should connect with client-side component code.

What about something more ambitious though? What about repeating lists? Well, another thing that's nice about Vue is that you can supply a template tag which defines a for loop.

```html
<template v-for="item in items">
  <li>{{ item.msg }}</li>
</template>
```

We could write this out something like:

```html
<template bindfor="item in items">
  <li bind="item.msg"></li>
</template>
```

Then the question becomes, how do you actually render out the list of items from the server, yet provide a mechanism to mutate that list on the client? I don't have a super great answer to that, but the thing I've prototyped in user land is basically this:

```html
<template bindfor="item in items">
  <li bind="item.msg"></li>
</template>
<li bindif="!hydrated">Message 1</li>
<li bindif="!hydrated">Message 2</li>
<li bindif="!hydrated">Message 3</li>
```

And once the client first processes these nodes, it essentially deletes the unhydrated nodes and replaces them with hydrated nodes…presumably with the exact same starting state so it's a relatively seamless process.

Does this cover all use cases? Not by any means. Yet it has the potential to solve quite a significant number of use-cases.

There's also outstanding possibilities like binding attributes (e.g., toggling a class on/off depending on a boolean) and all that, but I feel like that gets into the weeds of a spec. We first need to agree (a) do we need any spec at all, and (b) what basic problem would this new spec solve.

As I [wrote online](https://twitter.com/jaredcwhite/status/1568034913945001985?s=21&t=I_AUO_KEOy3Ogj3HZA-ASA):

> Right now the promise is this:
> 
> Write a component in JS and you can use it on both the server AND the client.
> 
> Or…you write a “component” in Ruby/Python/PHP/whatever but on the client it’s dumb. You need to write essentially a second component for some sort of hydration.

I really and truly love web components, but increasingly I find it's a very hard pill to swallow that in order to write any sort of "isomorphic" web component, I am unequivocally forced to drop all server architecture that isn't squarely based in JavaScript via Node/Deno/Bun/etc. That seems, [as I also wrote recently](https://twitter.com/jaredcwhite/status/1567982995960901636?s=21&t=I_AUO_KEOy3Ogj3HZA-ASA), to break a fundamental promise of web architecture: that the precise OS/language/runtime of the server is agnostic, and only protocols need to be agreed upon for client & server.

Thoughts?

**Reply by sashafirsov** [reply]
@jaredcwhite , this thread is wrong for proposal kind discussion. I would expect it on https://github.com/WICG/webcomponents/issues 
The data binding to slots and template in general is definitely nice to have in 5 years. Along with declarative data transformation and events/props/value binding..

Syntax of Parts with curly brackets is already in heavy discussion, kind of stock as of now. 
If you think it worth of discussion, lets create WICG stories and cross-refer here.

PS. I count the [Declarative Web Application](https://docs.google.com/document/d/1snAeYpAGMG9jP0z5tBv2XEe9hsIg84OyvnUYY77MUKM/edit?usp=sharing) stack definitely worth of pursue.

**Reply by trusktr** [reply]
Arguably `<output bind="count"></output>` is not as nice of a binding syntax as `<output>{{count}}</output>` for obvious reasons, but if the server uses the same interpolation syntax I see how that could get tricky.

What if we provide a way to configure custom interpolation markers? For example `bindstart="${" bindend="}"`:

```html
<template ... bindstart="${" bindend="}">
  <output>Name: ${name}</output>
</template>
```
or
```html
<template ... bindstart="#<" bindend=">#">
  <output>Name: #<name>#</output>
</template>
```

However, if the server side is generating data for a clientside template, maybe it is better to provide a way for a server to provide the initial values on the template element instead (and not provide syntax options to avoid complexity):

```html
<template ... init-first="John" init-last="Doe">
  <span>Name: {{first}} {{last}}</span>
</template>
```

Maybe even dataset is simply the existing bag of properties on which values live? That would align with what we already have:

```html
<template id="myTmpl" ... data-first-name="John" data-last-name="Doe">
  <span>Name: {{firstName}} {{lastName}}</span>
</template>

<script>
  // and then, reactivity on the client side:
  myTmpl.dataset.firstName = "Bat"
  myTmpl.dataset.lastName = "Man"
</script>
```

With this last idea of `data-` attributes and `dataset`, we already have the format for that, and the only thing we need to do is specify the template syntax, specify its behavior, give users an opt-in for templating (to make the template non-inert, if not already a shadowroot, either with an attribute such as `inlined` to make it render in line, or `myTmpl.createView()`), and finally let users interact with dataset for reactivity on inline templates (or `view.dataset` on view instances that take initial values from the template element's dataset).

Inlined (it renders to the screen, no manual template cloning or view creating):

```html
<template id="myTmpl" inlined data-first-name="John" data-last-name="Doe">
  <span>Name: {{firstName}} {{lastName}}</span>
</template>

<script>
  // and then, reactivity on the client side:
  myTmpl.dataset.firstName = "Bat"
  myTmpl.dataset.lastName = "Man"
</script>
```

Or manually create a view:

```html
<template id="myTmpl" data-first-name="John" data-last-name="Doe">
  <span>Name: {{firstName}} {{lastName}}</span>
</template>

<script>
  // and then, reactivity on the client side:
  const view = myTmpl.createView()
  document.body.append(view)
  view.dataset.firstName = "Bat"
  view.dataset.lastName = "Man"
</script>
```

I'm thinking that this is the entire templating API (not showing attribute bindings, and basic list binding such as `{{#each items as item}}{{/#each}}`.

The idea is to make something as simple as possiblle with the smallest possible spec (having Template Parts or DOM Parts is too much, and is really more of an implementation details, instead we can let browsers implement reactivity in their own ways, and we could expand on the spec later if we determine it would be useful, without blocking a sooner benefit to vanilla web API users).

For shadowroot templates it implies a single template view associated with the `shadowRoot`, and either `shadowRoot.dataset.foo = "..."` or `shadowRoot.view.dataset.foo = "..."` for reactivity (probably the latter for consistency, with `tmpl.dataset.*` being mapped to `tmpl.view.dataset.*` for convenience with inline templates).

Later on, the DCE spec could automatically map attribute/prop changes to template views (shadowroot or not).

Updates would be fine-grained. There would be no need for DOM/Template Parts, as the browser would handle the mapping internally, for simplicity of the spec.

Users would be able to fine-grained update:

```js
view.dataset.foo = 123
```

or bulk update:

```js
Object.assign(view.dataset, {bagOfNewValues})
```
